torch/csrc/jit/passes/quantization.cpp

```diff
@@ -0,0 +1,151 @@
+#include <cstdbool>
+#include <torch/csrc/jit/passes/quantization.h>
+
+#include <torch/csrc/jit/ir.h>
+#include <torch/csrc/jit/node_hashing.h>
+#include <torch/csrc/jit/passes/alias_analysis.h>
+
+#include <unordered_map>
+#include <unordered_set>
+#include <cstdint>
+
+namespace torch {
+namespace jit {
+namespace {
+// QuantizerUtils
+
+bool checkIfNodeValuesQuantizable(torch::jit::Node* n) {
+  // This is lookup for quantizable nodes. It will be
+  // expanded in future to support more ops and patterns
+  static const std::unordered_set<std::string> nodeLookup =
+    {"aten::_convolution", "aten::relu"};
+
+  if (n == nullptr) {
+    return false;
+  }
+
+  if (nodeLookup.find(n->schema().name()) != nodeLookup.end()) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+// Add Quant-Dequant node pair after quantizable node
+void addQuantDequantAfter(torch::jit::Value* v) {
+  torch::jit::Node* def = v->node();
+  AT_ASSERT(def != nullptr);
+  torch::jit::Node* quant =
+      def->owningGraph()->create(at::Symbol::fromQualString("aten::quant"));
+  AT_ASSERT(quant != nullptr);
+  torch::jit::Node* dequant =
+      def->owningGraph()->create(at::Symbol::fromQualString("aten::dequant"));
+  AT_ASSERT(dequant != nullptr);
+  def->owningBlock()->appendNode(quant);
+  def->owningBlock()->appendNode(dequant);
+  quant->setScope(def->scope());
+  dequant->setScope(def->scope());
+  quant->moveAfter(def);
+  dequant->moveAfter(quant);
+  v->replaceAllUsesWith(dequant->outputs()[0]);
+  quant->outputs()[0]->setUniqueName(v->uniqueName() + ".quant");
+  dequant->outputs()[0]->setUniqueName(v->uniqueName() + ".dequant");
+  quant->addInput(v);
+  dequant->addInput(quant->outputs()[0]);
+}
+
+// Add Quant-Dequant node pair before quantizable node
+void addQuantDeQuantBefore(torch::jit::Node* def, torch::jit::Value* v) {
```

Comment:
`addQuantDeQuantAfter` and `addQuantDeQuantBefore` are almost identical, and one of them is almost certainly not needed. Since the entire graph is
N -> V -> N -> V ...
Each V is both input of something and output of something (except values which are external inputs or outputs), and thus quantizing only inputs or only outputs should be sufficient.

Reply:
I had tried this approach. Consider this
N->(V1, V2)
V1 -> (N1, N2)
V2 -> (N3, N4)

Now i am interested in inserting quant node between V1->N2 and V2->N4. If i try doing the output only approach. i need to spot at N, scan over all outputs Vi, then scan over all uses of Vi and insert quant node. Also this require having some api like replaiceSpecificUseWith() which is not supported. Instead JIT provides replaceInputWith and replaceOutputWith which operates at node level. So if we are interested in V,N we use replaceInputWith and for N,V we use replaceOutputWith. Please let me know your thoughts

Reply:
What you described meant that you wanted to quantize inputs of N2 and N4.
What I tried to say is that if you go over all inputs of all nodes in the graph you will visit all values in the graph (except external outputs). Similarly, if you go through all outputs of all nodes, you will also visit all values in the graph (except external inputs). Thus, there is no reason to visit both inputs and outputs - in this case you will visit every value twice (except external inputs and outputs which would need a special handling anyway).

Reply:
Let me explain with input only case : 
we will visit only quantizable nodes. if we do just input or output, we will have to visit all nodes and also we might have more quant nodes than needed. Assuming we do inputs only approach, now for this case  Eg V0 -> N -> V1 ->(N1, N2, N3) will look like V0 -> Q -> N -> V1 -> (Q1,N1; Q2,N2; Q3,N3).  We cannot use replaceAll at input because of previous case of N2 and N4 that i described. If this case can be avoided then i agree only input approach is good

If we do both input and output, i agree we will have duplicate node visit in case N1->N2 are both quantizable, but graph will look like : V0 -> Q -> N -> V1 -> Q1->(N1; N2; N3)..

Reply:
Actually even for case  where consecutive nodes are quantizable, we will skip the second loop because there is a check for dequant node